processor/fanoutconnector.go

```diff
@@ -37,11 +43,24 @@ var _ MetricsProcessor = (*metricsFanOutConnector)(nil)
 // ConsumeMetricsData exports the MetricsData to all consumers wrapped by the current one.
 func (mfc metricsFanOutConnector) ConsumeMetricsData(ctx context.Context, md consumerdata.MetricsData) error {
 	var errs []error
-	for _, mc := range mfc {
-		if err := mc.ConsumeMetricsData(ctx, md); err != nil {
+
+	// Fan out to first len-1 consumers.
+	for i := 0; i < len(mfc)-1; i++ {
```

Comment:
Should this be an option in the "service" config for this? I think of cases where data are fanout but immutable?

Reply:
I think the only case where it would be safe is if the pipelines do not contain processors or exporters that modify data. I don't think it is safe to expose such option to end users.

What we can do is allow processors and exports to self-declare that they will not modify the data and then PipelineBuilder can check if currently active processors/exporters are only non-modifying kind then we can avoid cloning.

To support that we can add something like `GetCapabilities` function to processor and exporter interfaces.

Reply:
I agree that every processor should say immutable/copy_on_write/mutable or something like that, so then we know what we have to do. Maybe same applies to all the consumers?

Reply:
Yes, it should apply to all consumers, in our case we have 2 types of consumers: processors and exporters.

Reply:
> Maybe same applies to all the consumers?

Makes sense since processors and exporters share the same consumer interface.